### ISSUE TITLE
Use "in" at Matrix4x4

### ISSUE THREAD

**Issue by aobatact** [dotnet/corefx#25974]
It might be better to use in keyword at some methods of Matrix4x4

## Benchmark
``` csharp
static Matrix4x4 Add(Matrix4x4 value1, Matrix4x4 value2);
static Matrix4x4 InAdd(in Matrix4x4 value1,in Matrix4x4 value2);
static Matrix4x4 Multiply(Matrix4x4 value1, Matrix4x4 value2);
static Matrix4x4 InMultiply(in Matrix4x4 value1, in Matrix4x4 value2);
```
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.125)
Processor=Intel Core i7-7700K CPU 4.20GHz (Kaby Lake), ProcessorCount=8
.NET Core SDK=2.1.2
  [Host]     : .NET Core 2.0.3 (Framework 4.6.25815.02), 64bit RyuJIT  [AttachedDebugger]
  DefaultJob : .NET Core 2.0.3 (Framework 4.6.25815.02), 64bit RyuJIT


```
| Method |     Mean |     Error |    StdDev |
|------- |---------:|----------:|----------:|
|    Add | 14.01 ns | 0.1274 ns | 0.1130 ns |
|  InAdd | 13.11 ns | 0.0094 ns | 0.0088 ns |
|    Multiply | 30.34 ns | 0.0397 ns | 0.0287 ns |
|  InMultiply | 29.36 ns | 0.1664 ns | 0.1557 ns |

**Comment by karelz** [comment]
I assume the In-versions reuse the first argument, and return it as result. Is that correct?
Wouldn't it be better if it is instance method on the type then?
Given that we do not have them yet (not in [Numerics](https://apisof.net/catalog/System.Numerics)), I wonder if it is pattern we want to introduce.

Also I wonder if C# doesn't make `+=` work based off `+` operator (I heard something like it, but never dug deeper into it myself - check the generated IL).

BTW: `InMultiply` will need a scratch space anyway, right? I don't see how it can be faster than `Multiply`.

**Comment by tannergooding** [comment]
There was a discussion about some of these methods using in already and it was rejected in PR (on phone, so I can't find the link right now).

The end goal should be to fix any code generated issues here so that passing by value happens in register like  rather than implicitly by ref with shadow copying or by explicitly passing these by ref using in (passing by register without shadow copying is the default per the System V ABI on non-windows, and the default under the vectorcall ABI on Windows).

**Comment by tannergooding** [comment]
Cc. @eerhardt

**Comment by aobatact** [comment]
Both by value codes are copied from the corefx and by ref codes are just adding in so these do not use a first argument.
```csharp
static Matrix4x4 InAdd(in Matrix4x4 value1,in Matrix4x4 value2)
{
    Matrix4x4 m;

    m.M11 = value1.M11 + value2.M11;
    .../// from M12 to M43
    m.M44 = value1.M44 + value2.M44;

    return m;
}
static Matrix4x4 InMultiply(in Matrix4x4 value1, in Matrix4x4 value2)
{
    Matrix4x4 m;

    m.M11 = value1.M11 * value2.M11 + value1.M12 * value2.M21 + value1.M13 * value2.M31 + value1.M14 * value2.M41;
    ...//from M12 to M43
    m.M44 = value1.M41 * value2.M14 + value1.M42 * value2.M24 + value1.M43 * value2.M34 + value1.M44 * value2.M44;

    return m;
}
```

**Comment by karelz** [comment]
Oh, it just adds the `in` modifier on arguments. I didn't get it earlier.
That should be fairly straightforward and non-controversial change, assuming it really helps JIT to emit better code.

Can you show the difference in assembly generated by JIT between the 2 cases? (BenchmarkDotNet should have the capability)

**Comment by mikedn** [comment]
Here's the code generated by the JIT for the `Add` benchmark:
```asm
G_M16504_IG01:
       4881ECA8000000       sub      rsp, 168
       C5F877               vzeroupper
G_M16504_IG02:
       3909                 cmp      dword ptr [rcx], ecx
       488D9188000000       lea      rdx, bword ptr [rcx+136]
       4C8D4108             lea      r8, bword ptr [rcx+8]
       C4C17A6F00           vmovdqu  xmm0, qword ptr [r8] ; begin 64 byte copy
       C4E17A7F442468       vmovdqu  qword ptr [rsp+68H], xmm0
       C4C17A6F4010         vmovdqu  xmm0, qword ptr [r8+16]
       C4E17A7F442478       vmovdqu  qword ptr [rsp+78H], xmm0
       C4C17A6F4020         vmovdqu  xmm0, qword ptr [r8+32]
       C4E17A7F842488000000 vmovdqu  qword ptr [rsp+88H], xmm0
       C4C17A6F4030         vmovdqu  xmm0, qword ptr [r8+48]
       C4E17A7F842498000000 vmovdqu  qword ptr [rsp+98H], xmm0 ; end 64 byte copy
       4883C148             add      rcx, 72
       C4E17A6F01           vmovdqu  xmm0, qword ptr [rcx] ; begin 64 byte copy
       C4E17A7F442428       vmovdqu  qword ptr [rsp+28H], xmm0
       C4E17A6F4110         vmovdqu  xmm0, qword ptr [rcx+16]
       C4E17A7F442438       vmovdqu  qword ptr [rsp+38H], xmm0
       C4E17A6F4120         vmovdqu  xmm0, qword ptr [rcx+32]
       C4E17A7F442448       vmovdqu  qword ptr [rsp+48H], xmm0
       C4E17A6F4130         vmovdqu  xmm0, qword ptr [rcx+48]
       C4E17A7F442458       vmovdqu  qword ptr [rsp+58H], xmm0 ; begin 64 byte copy
       488BCA               mov      rcx, rdx
       488D542468           lea      rdx, bword ptr [rsp+68H]
       4C8D442428           lea      r8, bword ptr [rsp+28H]
       E847FAFFFF           call     System.Numerics.Matrix4x4:Add(struct,struct):struct
       90                   nop
G_M16504_IG03:
       4881C4A8000000       add      rsp, 168
       C3                   ret
```
There's also a 64 byte copy inside `Matrix4x4.Add`.

There are some cases where the JIT might be able to eliminate these copies but in this case (arguments are class fields) it's almost impossible. The JIT would need to look at `Matrix4x4.Add` and figure out that it doesn't in fact modify its arguments and thus copies are not needed.

That said, I'm not convinced that adding these overloads is really useful now, API clutter for less than 10% gain in performance. If `in` was available when `Matrix4x4` was first created perhaps doing this would have been a no brainer. Now the ship has kind of sailed.

Perhaps a better approach would be to investigate ways to communicate to the JIT that the method doesn't modify its arguments. Kind of ironically, adding `readonly` parameters rather than `in` parameters to C# might have been more useful in this regard.

**Comment by karelz** [comment]
@mikedn I agree that adding more APIs is not worth it. I assume we could just add `in` attribute on the args. I don't think that is a breaking change ...

**Comment by mikedn** [comment]
> I agree that adding more APIs is not worth it. I assume we could just add in attribute on the args. I don't think that is a breaking change ...

I don't think so, the resulting metadata signature would be incompatible with the existing signature so any C# code that uses `System.Numerics` would have to be recompiled.

**Comment by karelz** [comment]
If it is encoded as modopt in the signature, then yeah, we cannot just simply add it as it breaks binary compatibility. `out` and `ref` were not encoded that way. Is `in` different?

**Comment by mikedn** [comment]
`in` is like `ref`, the signature would contain a managed pointer:
```
void foo(valuetype [System.Numerics.Vectors]System.Numerics.Matrix4x4& m) cil managed
```
instead of the original one:
```
void foo(valuetype [System.Numerics.Vectors]System.Numerics.Matrix4x4 m) cil managed
```
Code that is not recompiled and attempts to load the new System.Numerics would fail with a `MissingMethodException`.

**Comment by tannergooding** [comment]
`in` has the same signature as `ref` and is differentiated by the compiler via the `S.R.CS.IsReadOnly` attribute.

I additionally uses a more for virtual members to guarantee the readonly contract when overriding. 

Since you cannot switch the methods to ref without it being breaking, you also cannot switch it to in.

Additionally, as commented above. Making many of these methods use in, or providing additional overloads to do so, was already rejected in a separate PR (cc. @eerhardt). I stead, it was decided to improve code generation where possible and ensure we followed the underlying abi which allows these to be passed in register without shadow copying (when passed by value).

**Comment by tannergooding** [comment]
The PR where this (and similar) change was already rejected is https://github.com/dotnet/corefx/pull/25388

**Comment by mikedn** [comment]
> Making many of these methods use in, or providing additional overloads to do so, was already rejected in a separate PR

The main problem with that PR was that it was trying do add `in` versions for many other `System.Numerics` types (e.g. `Vector3`) and for those types the benefits of using `in` are minor. In fact, some benchmarking showed that adding `in` can also make things slower sometimes.

The case of `Matrix4x4` is more complicated. It's much larger then the rest of the types and while the JIT may be able to eliminate some copies it's likely not possible to eliminate all copies. Especially on x86 where the calling convention is such that `by value` really means `by value`.

As such, there may still be value in adding such overloads to `Matrix4x4`.

**Comment by aobatact** [comment]
I came up with something like this
``` csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
static Matrix4x4 Add(Matrix4x4 value1,Matrix4x4 value2)
    => Add(in value1,in value2);

static Matrix4x4 Add(in Matrix4x4 value1,in Matrix4x4 value2);
```
but the user has to write "in" in the current way of overloading.
``` csharp
//var result = Add(matrix1,matrix2); CS0121
var result = Add(in matrix1,in matrix2);
```

**Comment by mikedn** [comment]
> but the user has to write "in" in the current way of overloading.

Yes, that's main reason why I said that the ship has kind of sailed. `in` has been specifically designed to be easier to use compared to `ref`, in normal circumstances you can call such methods without specifying `in`. But this only works if such methods aren't overloaded. Now we're sort of stuck:

* If such overloads are added now then existing code fails to compile.
* If C# overload resolution rules are "improved" to avoid the ambiguity then it's likely that the overload that doesn't use `in` will get selected (it makes more sense because it's safer to pass things by value instead of by reference). That means that developers who want to take advantage of `in` overloads will have to specify `in` and that kind of defeats the goal of `in` to enable simpler pass by reference.

Since there doesn't seem to be a way to add such overloads and avoid clumsy code we may as well go further and add `void Add(in Matrix4x4, in Matrix4x4, out Matrix4x4)`. Using `out` results in uglier code but it avoids the overload ambiguity problem and also avoids the struct copy implied by return by value.

**Comment by tannergooding** [comment]
@mikedn, the overload ambiguity is being resolved for 7.3.

Byval will be preferred when both exist and users will need to explicitly specify in to call the other overload.

https://github.com/dotnet/csharplang/issues/945
https://github.com/dotnet/roslyn/pull/23122

**Comment by mikedn** [comment]
> Byval will be preferred when both exist and users will need to explicitly specify in to call the other overload.

Exactly as I suspected it will happen :smile:
